rollbar/__init__.py

```diff
@@ -1606,3 +1628,34 @@ def _wsgi_extract_user_ip(environ):
     if real_ip:
         return real_ip
     return environ['REMOTE_ADDR']
+
+
+scopes = []
+
+
+class _BigBrother(object):
+    """
+    Context manager object that interfaces with the `scopes` stack:
+    
+        On enter, puts current scope object on top of stack.
+        On exit and there is no exception, pops off top element from stack.
+    """
+    def __init__(self, key, extra_data):
+        scope = {'key': key}
+
+        if extra_data:
+            extra_data = extra_data
+            if not isinstance(extra_data, dict):
+                extra_data = {'extra': extra_data}
+
+            scope = dict_merge(scope, extra_data, silence_errors=True)
+
+        self.scope = scope
+
+    def __enter__(self):
+        scopes.append(self.scope)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if not exc_type:
+            scopes.pop()
```

Comment:
It looks like if the user's code pattern catches exceptions outside the with statement, the scope will be leaked onto the stack and not recovered.

```
try:
  with Rollbar.watch('foobar'):
    raise Exception('any error')
except Exception as error:
  # Whether or not this is reported to Rollbar,
  # the 'foobar' scope isn't removed from scopes
```

Not sure if this is important for this release, or I might be misreading the intent for how this is used.

Reply:
Hm. Good catch! That's a really good point and definitely an unwanted consequence. I'd say this set up is alright, but I believe the desired intent while using the context manager is to wrap code where we want to report an exception, so one way to mitigate the leaked scope might be to just clear out `scopes` after a `report_exc_info()`.

Reply:
@ajtran I don't think we want to clear out `scopes` after report exception or message, it should happen in `__exit__` and always in `__exit__` no?

Outside `watch()` the scope added by `watch()` should be removed. If someone is reporting something inside `watch()` those scopes will be reported to Rollbar, but not if the exception is reported outside `watch()`.

Reply:
Ah. I see! Good catch. I believe you are correct. We can always clear out `scopes` in the `__exit__`, so that there is no `scope` leak if an exception does occur and if the user wants the scope with the exception, they should report it within the `watch()`

I'm wondering how this would look for a framework where we capture uncaught errors automatically tho i.e. django. I'll look into testing this tomorrow.

Reply:
Argh. So for frameworks where we capture uncaught errors automatically such as django, we aren't getting the data set by the context manager over to Rollbar because the error is reported outside of the context manager.

@jondeandres Can you explain how we can proceed by editing the exception, so the data isn't lost when the exception bubbles up?

Reply:
I guess you can use `setattr(exc, '__rollbar_scope', {'foo': 'bar'})`